src/hal/serial.cpp

```diff
@@ -0,0 +1,137 @@
+
+// this is the storage for getting bytes Serial.readBytes(1), reading bytes one at a time. probably could use pointer, but used union command bytes to get one byte per call
+// when the full packet is received this is put copied to the public command data structure.
+typedef union command_packet_union{
+command_packet_def command_packet;
+uint8_t command_bytes[sizeof(command_packet_def)];
+};
+
+command_packet_union command_packet_u;
+
+int serialHalInit(void)
+{
+  serial_ui.begin(BAUD_RATE);  // this function does not return anything
+  // not sure how we can do something like while (!serial)
+  // return MODULE_FAIL;
+  do
+  { serial_ui.read();
+  } while (serial_ui.available() > 0);
+  
+#ifdef SERIAL_DEBUG
+  serial_debug.begin(BAUD_RATE);
+#endif  
+  return HAL_OK;
+}
+
+int serialHalGetData(void)
+{
+  // change to while if reading all the bytes currently available
+  if (serial_ui.available())
+  {       
+    serial_ui.readBytes(&inByte, 1);
+    
+    command_packet_u.command_bytes[incoming_index] = inByte;
+    
+    incoming_index++;
+
+    if (incoming_index >= sizeof(comm.public_command_packet))
+    {
+      // save a copy for other modules, but keep a reference in case the shared copy gets modified
+      incoming_index = 0;
+      watchdog_active = false;
+
+      memcpy((void *)&comm.public_command_packet, (void *)&command_packet_u.command_packet, sizeof(comm.public_command_packet));
+      // clear alarm bits
+      comm.public_command_packet.alarm_bits &= ~(1 << ALARM_DROPPED_PACKET);
+#ifdef SERIAL_DEBUG
+      serial_debug.print("Received from Rpi: 0x");
+      serial_debug.print(comm.public_command_packet.sequence_count, HEX);
+      serial_debug.print(" 0x");
+      serial_debug.println(comm.public_command_packet.crc, HEX);
+      serial_debug.println(" ");
+#endif      
+      return HAL_OK;
+    } 
+  }
+  if (watchdog_active == true)
+  {
+    if ((millis() - watchdog_start_ms) > watchdog_max_ms)
+    {
+#ifdef SERIAL_DEBUG
+      serial_debug.println("!!!Watchdog timeout");
+#endif        
+      watchdog_active = false; 
+      watchdog_exceeded = true; 
+      incoming_index = 0;
+      return HAL_OK;        
+    }
+  } // if serial.available
+  return HAL_IN_PROGRESS;
+}
+
+int serialHalSendData()
+{
+  current_send_ms = millis();
+  if ((current_send_ms - last_send_ms) >= send_interval_ms)
+  {
+    if (clear_input == true)
+    {
+      clear_input = false;
+      do
+      { serial_ui.read();
+      } while (serial_ui.available() > 0);
+    }
+#ifdef SERIAL_DEBUG
+    serial_debug.print("Sent to Rpi: ");
+    serial_debug.print(comm.public_data_packet.sequence_count, HEX);
+    serial_debug.print(" ");
+    serial_debug.print(comm.public_data_packet.sequence_count, HEX);
+    serial_debug.println(" ");
+#endif 
+       
+    bytesSent = serial_ui.write((byte *)&comm.public_data_packet, sizeof(comm.public_data_packet));
```

Comment:
The Serial.write() function returns before any byte is sent unless the data to write is longer than the allocated buffer size. It will block for the excess bytes. We have at least 70 bytes in the data packet so we will need to increase this size in HardwareSerial.h so that we are not blocking.
#if !defined(SERIAL_TX_BUFFER_SIZE)
#if ((RAMEND - RAMSTART) < 1023)
#define SERIAL_TX_BUFFER_SIZE 16
#else
#define SERIAL_TX_BUFFER_SIZE 64
#endif
#endif

there are other ways to do this 
https://www.hobbytronics.co.uk/arduino-serial-buffer-size